src/lem/store_core.rs

```diff
@@ -0,0 +1,398 @@
+use arc_swap::ArcSwap;
+use elsa::sync::{index_set::FrozenIndexSet, FrozenMap, FrozenVec};
+use indexmap::IndexSet;
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use std::sync::Arc;
+
+use super::pointers::{GPtr, IPtr, IVal};
+
+/// Trait that teaches the `StoreCore` how to perform custom hashing operations
+pub trait StoreHasher<T, D> {
+    fn hash_ptrs(&self, ptrs: Vec<GPtr<T, D>>) -> D;
+    fn hash_compact(&self, d1: D, t2: T, d2: D, d3: D) -> D;
+    fn hash_commitment(&self, secret: D, payload: GPtr<T, D>) -> D;
+}
+
+/// A data structure used to efficiently encode data as DAGs of tagged pointers
+/// that can eventually be content-addressed by a custom hasher
+#[derive(Debug)]
+pub struct StoreCore<T, D, H: StoreHasher<T, D>> {
+    /// Holds leaf (non-compound) data
+    atom: FrozenIndexSet<Box<D>>,
+
+    /// Holds compound data with 2 children
+    tuple2: FrozenIndexSet<Box<[IPtr<T>; 2]>>,
+
+    /// Holds compound data with 3 children
+    tuple3: FrozenIndexSet<Box<[IPtr<T>; 3]>>,
+
+    /// Holds compound data with 4 children
+    tuple4: FrozenIndexSet<Box<[IPtr<T>; 4]>>,
+
+    /// `StoreHasher` implementer
+    pub hasher: H,
+
+    /// Holds the commitments performed with the store
+    pub comms: FrozenMap<D, Box<(D, IPtr<T>)>>, // digest -> (secret, payload)
+
+    /// Keeps track of data that hasn't been hashed yet for future parallel hashing
+    dehydrated: ArcSwap<FrozenVec<Box<IVal>>>,
+
+    /// Memoized hashes
+    z_cache: FrozenMap<IVal, Box<D>>,
+
+    /// Inverse hashes to enable the retrieval of the original data from their
+    /// digests
+    inverse_z_cache: FrozenMap<D, Box<IVal>>,
+}
+
+impl<
+        T: PartialEq + std::cmp::Eq + std::hash::Hash,
+        D: PartialEq + std::cmp::Eq + std::hash::Hash,
+        H: StoreHasher<T, D> + Default,
+    > Default for StoreCore<T, D, H>
+{
+    fn default() -> Self {
+        Self {
+            atom: Default::default(),
+            tuple2: Default::default(),
+            tuple3: Default::default(),
+            tuple4: Default::default(),
+            hasher: Default::default(),
+            comms: Default::default(),
+            dehydrated: Default::default(),
+            z_cache: Default::default(),
+            inverse_z_cache: Default::default(),
+        }
+    }
```

Comment:
Can't you derive this?

Reply:
I don't think so because it requires further implementations for `T` and `D`